modules/diskq/diskq-global-metrics.c

```diff
@@ -0,0 +1,197 @@
+{
+  struct statvfs stat;
+  if (statvfs(dir, &stat) < 0 )
+    {
+      msg_error("disk-buffer: Failed to get filesystem info",
+                evt_tag_str("dir", dir),
+                evt_tag_error("error"));
+      return FALSE;
+    }
+
+  *available_space_mib = stat.f_bsize * stat.f_bavail / (1024 * 1024);
+
+  return TRUE;
+}
+
+static void
+_update_dir_metrics(gpointer key, gpointer value, gpointer user_data)
+{
+  const gchar *dir = (const gchar *) key;
+
+  gint64 available_space_mib;
+  if (!_get_available_space_mib_in_dir(dir, &available_space_mib))
+    return;
+
+  StatsClusterLabel labels[] = { stats_cluster_label("dir", dir), };
+  StatsClusterKey sc_key;
+  stats_cluster_single_key_set(&sc_key, "disk_buffer_dir_available_bytes", labels, G_N_ELEMENTS(labels));
+  /* Up to 4096 TiB with 32 bit atomic counters. */
+  stats_cluster_single_key_add_unit(&sc_key, SCU_MIB);
+
+  stats_lock();
+  {
+    StatsCounterItem *counter;
+    StatsCluster *cluster = stats_register_dynamic_counter(STATS_LEVEL1, &sc_key, SC_TYPE_SINGLE_VALUE, &counter);
+    stats_counter_set(counter, available_space_mib);
+    stats_unregister_dynamic_counter(cluster, SC_TYPE_SINGLE_VALUE, &counter);
+  }
+  stats_unlock();
+}
+
+static void
+_update_all_dir_metrics(gpointer s)
+{
+  DiskQGlobalMetrics *self = (DiskQGlobalMetrics *) s;
+
+  g_mutex_lock(&self->lock);
+  {
+    g_hash_table_foreach(self->dirs, _update_dir_metrics, NULL);
+  }
+  g_mutex_unlock(&self->lock);
+
+  _dir_watch_timer_start(self);
+}
+
+static void
+_new(gint type, gpointer c)
+{
+  DiskQGlobalMetrics *self = &diskq_global_metrics;
+
+  IV_TIMER_INIT(&self->dir_watch_timer);
+  self->dir_watch_timer.handler = _update_all_dir_metrics;
+  self->dir_watch_timer.cookie = self;
+
+  self->dirs = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+}
+
+static void
+_init(gint type, gpointer c)
+{
+  GlobalConfig *cfg = main_loop_get_current_config(main_loop_get_instance());
+  if (!cfg)
+    return;
+
+  DiskQGlobalMetrics *self = &diskq_global_metrics;
+
+  self->freq = disk_queue_config_get_stats_freq(cfg);
+  if (self->freq == 0)
+    return;
+
+  _update_all_dir_metrics(self);
+}
+
+static void
+_deinit(gint type, gpointer c)
+{
+  DiskQGlobalMetrics *self = &diskq_global_metrics;
+
+  _dir_watch_timer_stop(self);
+
+  g_mutex_lock(&self->lock);
+  {
+    g_hash_table_remove_all(self->dirs);
+  }
+  g_mutex_unlock(&self->lock);
+}
+
+static void
+_free(gint type, gpointer c)
+{
+  DiskQGlobalMetrics *self = &diskq_global_metrics;
+
+  g_hash_table_destroy(self->dirs);
+}
+
+void
+diskq_global_metrics_init(void)
+{
+  DiskQGlobalMetrics *self = &diskq_global_metrics;
+
+  if (self->global_inited)
+    return;
+
+  register_application_hook(AH_STARTUP, _new, NULL, AHM_RUN_ONCE);
+  register_application_hook(AH_CONFIG_CHANGED, _init, NULL, AHM_RUN_REPEAT);
+  register_application_hook(AH_CONFIG_STOPPED, _deinit, NULL, AHM_RUN_REPEAT);
```

Comment:
I double-checked, and it is not necessary to make _init and _deinit idempotent as these hooks "work correctly" when the config is reverted.

Optional note:
_deinit() is already idempotent, we could make _init() too if we protected the `_update_all_dir_metrics()` call.

Reply:
Reregistering an existing timer will cause a crash.
I just wanted to let you know about this, but I'm okay with not fixing this.

Reply:
I would like to prepare this for darker times :D Fixed.

Reply:
Thank you.